src/storage-client/src/source/persist_source.rs

```diff
@@ -292,96 +292,94 @@ where
-            let mut inflight_bytes = 0;
-            let mut inflight_parts: Vec<(Antichain<Timestamp>, usize)> = Vec::new();
-
-            move |cap_set, flow_control_upper, output| {
-                let mut context = Context::from_waker(&waker);
-
-                let retired_parts = inflight_parts.drain_filter_swapping(|(upper, _size)| {
-                    PartialOrder::less_equal(&upper.borrow(), &flow_control_upper)
-                });
-
-                for (_upper, size_in_bytes) in retired_parts {
-                    inflight_bytes -= size_in_bytes;
-                    trace!(
-                        "shard {} returning {} bytes. total: {}. batch frontier {:?} less_equal to {:?}",
-                        data_shard,
-                        size_in_bytes,
-                        inflight_bytes,
-                        _upper,
-                        flow_control_upper,
-                    );
-                }
+        scope.clone(),
+    );
+    let flow_control_input = builder_dist.new_input(flow_control_input, Pipeline);
+    let (mut parts_output, parts_stream) = builder_dist.new_output();
+    let shutdown_button = builder_dist.build(move |caps| async move {
+        let mut cap_set = CapabilitySet::from_elem(caps.into_element());
 
-                while inflight_bytes < flow_control_max_inflight_bytes {
-                    match pinned_stream.as_mut().poll_next(&mut context) {
-                        Poll::Ready(Some(Ok((parts, progress, size_in_bytes)))) => {
-                            let session_cap = cap_set.delayed(&current_ts);
-                            let mut session = output.session(&session_cap);
-
-                            if size_in_bytes > 0 {
-                                inflight_parts.push((progress.clone(), size_in_bytes));
-                                inflight_bytes += size_in_bytes;
-                                trace!(
-                                    "shard {} putting {} bytes inflight. total: {}. batch frontier {:?}",
-                                    data_shard,
-                                    size_in_bytes,
-                                    inflight_bytes,
-                                    progress,
-                                );
-                            }
+        let mut current_ts = timely::progress::Timestamp::minimum();
+        let mut inflight_bytes = 0;
+        let mut inflight_parts: Vec<(Antichain<Timestamp>, usize)> = Vec::new();
 
-                            for part in parts {
-                                // Give the part to a random worker.
-                                let worker_idx = usize::cast_from(Instant::now().hashed()) % peers;
-                                session.give((worker_idx, part.into_exchangeable_part()));
-                            }
+        while inflight_bytes < flow_control_max_inflight_bytes {
+            let flow_control_upper = flow_control_input.frontier();
```

Comment:
@petrosagg any advice for how I should be getting this frontier? what I have in the PR doesn't feel right but it wasn't immediately obvious to me what I should be doing instead. maybe using `select!` over both `flow_control_input.next()` and `pinned_stream.next()`?

Reply:
I think you can avoid `select!` but you definitely want to be getting it through an await call so that the future gets waken when the frontier makes progress. I think something like this will do the trick:


```rust
while inflight_bytes < flow_control_max_inflight_bytes {
    // Get an upper bound until which we should produce data
    let flow_control_upper = loop {
        if let Event::Progress(frontier) = flow_control_input.next().await {
            break frontier;
        }
    };

    // retire parts
    let retired_parts = inflight_parts.drain_filter_swapping(|(upper, _size)| {
        PartialOrder::less_equal(&*upper, &flow_control_upper)
    });

    // schedule more work according to the flow control upper
    while PartialOrder::less_equal(&flow_control_upper, &*upper) {
        match pinned_stream.next().await {
            ...
        }
    }
}
```

The code above has the problem that there is no notification for the frontier being at minimum which sounds like a problem. I think we can change the default value of the `pending` flag to be `true` so that we always give a progress event at start up. It's set here https://github.com/MaterializeInc/materialize/blob/main/src/timely-util/src/builder_async.rs#L230

Reply:
I forgot the byte check above but I think the general structure still stands (sorry I've got super slow airplane wifi)

Reply:
yeah okay, I see!

I was having trouble convincing myself that we'd be guaranteed to stay reactive to new subscription batches showing up with the flow_control_upper based while loops, what about this idea I just pushed that does something similar, but bases the while loops on inflight_bytes vs flow_control_max_inflight_bytes?

the structure is a bit hard to see because some of the loops are kinda long, but it's roughtly
```
loop {
    while inflight_bytes < flow_control_max_inflight_bytes {
        // read batches from subscription and emit them
    }
    while inflight_bytes >= flow_control_max_inflight_bytes {
        // read frontier progress from flow_control_input and retire inflight batches
    }
}
```

one worry I have with that is e.g. if flow_control_max_inflight_bytes is usize::MAX and we never read from flow_control_input, does that back up a buffer in timely somewhere? we never write data to flow_control_input, so my wild guess is it could go either way

Reply:
That looks great, and also doesn't rely on having an initial Progress message. Nice! Data would be buffering in the handle if it was transferring data but since we don't there isn't any risk (progress `ChangeBatches` are consumed even if you don't interact with the handle. Ideally the stream should be a `Stream<G, !>` to statically ensure that no data can ever be sent

Reply:
Made the stream `Infallible` which seems like a decent holdover until we have `!`. Also added your comment here as commentary in case some one else reads this later and gets stressed.

Okay, I think this should be ready to go!